Design.md

```diff
@@ -0,0 +1,149 @@
+
+The job-runner uses cgroups V1 for resource limits. Cgroups will be used through an API wrapper like [controlgroup-rs](https://github.com/ordovicia/controlgroup-rs) or [cgroups-fs](https://github.com/frol/cgroups-fs). There are many tweakable options for limiting resource use, but the following minimal set has been selected to demonstrate how they work. In particular, the following cgroup parameters will be set:
+* [CPU](https://kernel.googlesource.com/pub/scm/linux/kernel/git/glommer/memcg/+/cpu_stat/Documentation/cgroups/cpu.txt):
+    * `cpu.cfs_period_us` (keep default: 100ms)
+    * `cpu.cfs_quota_us`
+        * `cfs_period_us` means one cpu cores worth of usage
+        * `cfs_period_us` / 2 means 50% of one core
+* [Memory](https://kernel.googlesource.com/pub/scm/linux/kernel/git/glommer/memcg/+/cpu_stat/Documentation/cgroups/memory.txt):
+    * `memory.limit_in_bytes` (max amount of memory to be used, hard limit)
+    * `memory.soft_limit_in_bytes` (set to 75% of `memory.limit_in_bytes`)
+    * `memory.memsw.limit_in_bytes` (set to `memory.limit_in_bytes`)
+* [Block IO](https://kernel.googlesource.com/pub/scm/linux/kernel/git/glommer/memcg/+/cpu_stat/Documentation/cgroups/blkio-controller.txt): (Device list readable from `/proc/partitions`)
+    * `blkio.throttle.read_bps_device`
+    * `blkio.throttle.write_bps_device`
+    * `blkio.throttle.read_iops_device`
+    * `blkio.throttle.write_iops_device`
+
+In addition to limiting resource use with cgroups, job-runner also isolates jobs from each other using namespaces. A PID namespace is set up to make sure the job cannot kill processes not spawned by it, and to make sure all child processes are terminated together with the actual job. A mount namespace is used to limit process to a subset of the file system, together with [`pivot_root(2)`](https://linux.die.net/man/2/pivot_root) (see [Understanding Containerization By Recreating Docker](https://itnext.io/linux-container-from-scratch-339c3ba0411d), search for pivot_root). Finally, a network namespace is created to limit network access of the jobs. It must be used together with [Virtual ethernet (VETH)](https://developers.redhat.com/blog/2018/10/22/introduction-to-linux-interfaces-for-virtual-networking#veth) if jobs should be allowed to communicate between each other or send data to the internet.
+
+## Communication and auth
+
+Clients communicate via gRPC with a simple protocol: One service with four endpoints. The communication is secured [RusTLS](https://github.com/ctz/rustls) which by design supports only modern, secure cipher suites. Only TLS 1.2/1.3 are used, and authentication is done using ECDSA, Ed25519 or RSA. RusTLS will remove support for cipher suites deemed insecure, and simply keeping the version up to date should be sufficient in the future as well. [Tonic](https://github.com/hyperium/tonic) is used to serve gRPC and almost automatically handles protocol buffers, encryption and related concerns.
```

Comment:
How do you know all the cipher suites chosen by rustTLS are secure?

Reply:
I don't, and I'm not a professional cryptography researcher so I cannot actually verify it myself. I did the preliminary analysis about the cipher suite selection, and the supported alternatives are widely regarded as solid options. Reasons why I trust RusTLS:

* ISRG (which runs Let's Encrypt) [praises it](https://www.abetterinternet.org/post/preparing-rustls-for-wider-adoption/) 
* Cure53 completed an [audit](https://github.com/ctz/rustls/blob/master/audit/TLS-01-report.pdf) on RusTLS about a year ago, and found no major issues
* It's used by actual widely-used software, like [curl](https://daniel.haxx.se/blog/2021/02/09/curl-supports-rustls/)
* The primitives are mostly same that are used by NaCL, which is considered the golden standard of cryptography libraries
* I somewhat trust the authors, as they have long history on maintaining open source and contributing to the cryptography software